sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +188,309 @@ protected AttestationToken()
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // We don't have a desired Key ID, so we need to try all the possible signing certificates :(.
+
+                // Start with the attestationSigners provided by the caller. Remember that attestation signers
+                // trump all signers in the header.
+                if (attestationSigners != null)
+                {
+                    // Copy the signers to the candidate certificates.
+                    //
+                    // Note: because we don't have a KeyID, we neuter the KeyID in the AttestationSigner.
+                    foreach (var signer in attestationSigners)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(signer.SigningCertificates.ToArray(), null));
+                    }
+                }
+                else
+                {
+                    if (Header.X509CertificateChain != null)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(ConvertBase64CertificateArrayToCertificateChain(Header.X509CertificateChain), null));
+                    }
+                    if (Header.JsonWebKey != null)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(ConvertBase64CertificateArrayToCertificateChain(Header.JsonWebKey.X5C), null));
+                    }
+                }
+            }
+            await Task.Yield();
+            return candidateCertificates.ToArray();
+        }
+
+        private static X509Certificate2[] ConvertBase64CertificateArrayToCertificateChain(IList<string> base64certificates)
+        {
+            List<X509Certificate2> jwkCertificates = new List<X509Certificate2>();
+            foreach (var base64Cert in base64certificates)
+            {
+                jwkCertificates.Add(new X509Certificate2(Convert.FromBase64String(base64Cert)));
+            }
+            return jwkCertificates.ToArray();
+        }
+
+        /// <summary>
+        /// Iterate over the set of possible signers looking for a signer which can validate the signature
+        /// </summary>
+        /// <param name="possibleSigners">A set of attestation signers which might have been used to sign the token.</param>
+        /// <returns>true if one of the possibleSigners signed the token, false otherwise</returns>
+        private bool ValidateTokenSignature(AttestationSigner[] possibleSigners)
+        {
+            bool signatureValidated = false;
+            AttestationSigner actualSigner = null;
+
+            foreach (var signer in possibleSigners)
+            {
+                // The leaf certificate is defined as the certificate which signed the token, so we just need to look
+                // at the first certificate in the chain.
+                AsymmetricAlgorithm asymmetricAlgorithm = signer.SigningCertificates[0].PublicKey.Key;
```

Comment:
Yes - the AsymmetricAlgorithm is either in the signingKey.Certificate or the signingKey.Signer. And that's a value which comes from the customer, so disposing it would be incorrect.